machine/v1/types_nutanixprovider.go

```diff
@@ -154,6 +165,156 @@ type NutanixResourceIdentifier struct {
 	Name *string `json:"name,omitempty"`
 }
 
+// NutanixGPUIdentifierType is an enumeration of different resource identifier types for GPU entities.
+// +kubebuilder:validation:Enum:=Name;DeviceID
+type NutanixGPUIdentifierType string
+
+const (
+	// NutanixGPUIdentifierName identifies a GPU by Name.
+	NutanixGPUIdentifierName NutanixGPUIdentifierType = "Name"
+
+	// NutanixGPUIdentifierDeviceID identifies a GPU by device ID.
+	NutanixGPUIdentifierDeviceID NutanixGPUIdentifierType = "DeviceID"
+)
+
+// NutanixGPU holds the identity of a Nutanix GPU resource in the Prism Central
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'DeviceID' ?  has(self.deviceID) : !has(self.deviceID)",message="deviceID configuration is required when type is DeviceID, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Name' ?  has(self.name) : !has(self.name)",message="name configuration is required when type is Name, and forbidden otherwise"
+// +union
+type NutanixGPU struct {
+	// type is the identifier type of the GPU device.
+	// Valid values are Name and DeviceID.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Type NutanixGPUIdentifierType `json:"type"`
+
+	// deviceID is the GPU device ID with the integer value.
+	// +optional
+	// +unionMember
+	DeviceID *int32 `json:"deviceID,omitempty"`
+
+	// name is the GPU device name
+	// +optional
+	// +unionMember
+	Name *string `json:"name,omitempty"`
+}
+
+// NutanixDiskMode is an enumeration of different disk modes.
+// +kubebuilder:validation:Enum=Standard;Flash
+type NutanixDiskMode string
+
+const (
+	// NutanixDiskModeStandard represents the disk standard mode (not flash).
+	NutanixDiskModeStandard NutanixDiskMode = "Standard"
+
+	// NutanixDiskModeFlash represents the disk flash mode.
+	NutanixDiskModeFlash NutanixDiskMode = "Flash"
+)
+
+// NutanixStorageResourceIdentifier holds the identity of a Nutanix storage resource (storage_container, etc.)
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'uuid' ?  has(self.uuid) : !has(self.uuid)",message="uuid configuration is required when type is uuid, and forbidden otherwise"
+// +union
+type NutanixStorageResourceIdentifier struct {
+	// Type is the identifier type to use for this resource.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=uuid
+	Type NutanixIdentifierType `json:"type"`
+
+	// uuid is the UUID of the storage resource in the PC.
+	// +optional
+	// +unionMember
+	UUID *string `json:"uuid,omitempty"`
+}
+
+// NutanixVMStorageConfig specifies the storage configuration parameters for VM disks.
+type NutanixVMStorageConfig struct {
+	// diskMode specifies the disk mode.
+	// The valid values are Standard and Flash, and the default is Standard.
+	// +kubebuilder:default=Standard
+	DiskMode NutanixDiskMode `json:"diskMode"`
```

Comment:
Perhaps it makes sense to call this something akin to "DiskTierPreference". By default, you can use up to 25% of the cluster-wide SSD tier as flash mode space for VMs or VGs. If the data size for flash mode enabled VMs or VGs exceeds 25% of the SSD capacity, the system may down migrate the data. 

See https://portal.nutanix.com/page/documents/details?targetId=Web-Console-Guide-Prism-v6_8:wc-vm-flash-mode-c.html

Reply:
I think it's okay to use the diskMode here.

Reply:
Is there a way to disable this downgrade behaviour? Requesting something and not getting it is very odd in Kube APIs, I'd have expected an error to be returned